snap/info_snap_yaml_test.go

```diff
@@ -39,14 +41,801 @@ type: app
 `)
 
 func (s *InfoSnapYamlTestSuite) TestSimple(c *C) {
-	info, err := InfoFromSnapYaml(mockYaml)
+	info, err := snap.InfoFromSnapYaml(mockYaml)
 	c.Assert(err, IsNil)
 	c.Assert(info.Name, Equals, "foo")
 	c.Assert(info.Version, Equals, "1.0")
-	c.Assert(info.Type, Equals, TypeApp)
+	c.Assert(info.Type, Equals, snap.TypeApp)
 }
 
 func (s *InfoSnapYamlTestSuite) TestFail(c *C) {
-	_, err := InfoFromSnapYaml([]byte("random-crap"))
+	_, err := snap.InfoFromSnapYaml([]byte("random-crap"))
 	c.Assert(err, ErrorMatches, "(?m)info failed to parse:.*")
 }
+
+type YamlSuite struct{}
+
+var _ = Suite(&YamlSuite{})
+
+func (s *YamlSuite) TestUnmarshalGarbage(c *C) {
+	_, err := snap.InfoFromSnapYaml([]byte(`"`))
+	c.Assert(err, ErrorMatches, ".*: yaml: found unexpected end of stream")
+}
+
+func (s *YamlSuite) TestUnmarshalEmpty(c *C) {
+	info, err := snap.InfoFromSnapYaml([]byte(``))
+	c.Assert(err, IsNil)
+	c.Assert(info.Plugs, HasLen, 0)
+	c.Assert(info.Slots, HasLen, 0)
+	c.Assert(info.Apps, HasLen, 0)
+}
+
+// Tests focusing on plugs
+
+func (s *YamlSuite) TestUnmarshalStandaloneImplicitPlug(c *C) {
+	// NOTE: yaml content cannot use tabs, indent the section with spaces.
+	info, err := snap.InfoFromSnapYaml([]byte(`
+name: snap
+plugs:
+    network-client:
+`))
+	c.Assert(err, IsNil)
+	c.Check(info.Name, Equals, "snap")
+	c.Check(info.Plugs, HasLen, 1)
+	c.Check(info.Slots, HasLen, 0)
+	plug := info.Plugs["network-client"]
+
+	c.Assert(plug, Not(IsNil))
```

Comment:
I am not super happy about the amount of similar test code in this and the following tests. There is a lot of repetition (DRY violation) that does not feel great. Unfortunately I don't have a great idea to avoid it, I was wondering if:
```
c.Assert(plug, DeepEquals, &snap.PlugInfo{
		Snap:      info,
		Name:      "network-client",
		Interface: "network-client",
	})
```
would help. This would test the defaults (like plug.Attrs, HasLen, 0) implicitly and concentrate on the key point of the test. But in order for this to be "nice" the code should probably tweaked a little bit so that "snap.PlugInfo.Apps" is nil if empty instead of "map[string]*AppInfo{}" (i.e. the above will not work because of this difference). However I'm not sure if it is really worth it given that this yaml will hopefully never change in an incompatible way in 16.04. But I still feel a bit uneasy, it appears the tests have more identical lines than different lines which makes reading them slower because its less easy to see what the "key point" of the test is (i.e. what it tests differently from the previous test). But don't feel blocked, the level of testing is really nice.